pkg/registry/core/bastion/strategy.go

```diff
@@ -0,0 +1,189 @@
+// Copyright (c) 2021 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bastion
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/gardener/gardener/pkg/api"
+	"github.com/gardener/gardener/pkg/apis/core"
+	v1alpha1constants "github.com/gardener/gardener/pkg/apis/core/v1alpha1/constants"
+	"github.com/gardener/gardener/pkg/apis/core/validation"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/apiserver/pkg/registry/generic"
+	"k8s.io/apiserver/pkg/storage"
+	"k8s.io/apiserver/pkg/storage/names"
+)
+
+const (
+	// BastionTimeToLive is the duration the ExpirationTimestamp of a bastion will be extended
+	// by on every heartbeat.
+	BastionTimeToLive = 1 * time.Hour
+)
+
+type bastionStrategy struct {
+	runtime.ObjectTyper
+	names.NameGenerator
+
+	timeToLive time.Duration
+}
+
+// Strategy defines the storage strategy for Bastions.
+var Strategy = bastionStrategy{api.Scheme, names.SimpleNameGenerator, BastionTimeToLive}
+
+func (bastionStrategy) NamespaceScoped() bool {
+	return true
+}
+
+func (s bastionStrategy) PrepareForCreate(ctx context.Context, obj runtime.Object) {
+	bastion := obj.(*core.Bastion)
+
+	// at this point, CreationTimestamp is not yet set on the bastion
+	now := metav1.NewTime(time.Now())
+	expires := metav1.NewTime(now.Add(s.timeToLive))
+
+	bastion.Generation = 1
+	bastion.Status = core.BastionStatus{
+		LastHeartbeatTimestamp: &now,
+		ExpirationTimestamp:    &expires,
+	}
+}
+
+func (bastionStrategy) PrepareForUpdate(ctx context.Context, obj, old runtime.Object) {
+	newBastion := obj.(*core.Bastion)
+	oldBastion := old.(*core.Bastion)
+	newBastion.Status = oldBastion.Status
+
+	if mustIncreaseGeneration(oldBastion, newBastion) {
+		newBastion.Generation = oldBastion.Generation + 1
+	}
+}
```

Comment:
What about the following part outlined in the [GEP](https://github.com/gardener/gardener/blob/master/docs/proposals/15-manage-bastions-and-ssh-key-pair-rotation.md):

> when gardener.cloud/operation: keepalive is set it will be removed by GAPI from the annotations and status.lastHeartbeatTimestamp will be set with the current timestamp. The status.expirationTimestamp will be calculated by taking the last heartbeat timestamp and adding x minutes (configurable, default 60 Minutes).

This needs to be added to the `PrepareForUpdate` function.